elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,247 @@
+ *
+ * FilterExpression whereClauseFilter = filterPair.getLeft();
+ * FilterExpression havingClauseFilter = filterPair.getRight();
+ * }
+ * </pre>
+ * {@link SplitFilterExpressionVisitor} is thread-safe and can be accessed by multiple threads at the same time.
+ */
+@Slf4j
+public class SplitFilterExpressionVisitor implements FilterExpressionVisitor<Pair<FilterExpression, FilterExpression>> {
+
+    @Getter(value = AccessLevel.PRIVATE)
+    private final EntityDictionary entityDictionary;
+    @Getter(value = AccessLevel.PRIVATE)
+    private final FilterExpressionNormalizationVisitor normalizationVisitor;
+
+    /**
+     * Constructor.
+     *
+     * @param entityDictionary  Object that offers annotation information about an entity field
+     * @param normalizationVisitor  A helper {@link FilterExpressionVisitor} that is leveraged to transform
+     * expressions like {@code NOT (A AND B)} into {@code NOT A OR NOT B}
+     */
+    public SplitFilterExpressionVisitor(
+            final EntityDictionary entityDictionary,
+            final FilterExpressionNormalizationVisitor normalizationVisitor
+    ) {
+        this.entityDictionary = entityDictionary;
+        this.normalizationVisitor = normalizationVisitor;
+    }
+
+    @Override
+    public Pair<FilterExpression, FilterExpression> visitPredicate(final FilterPredicate filterPredicate) {
+        return isHavingPredicate(filterPredicate)
+                ? Pair.of(null, filterPredicate)   // this filterPredicate belongs to a HAVING clause
+                : Pair.of(filterPredicate, null); // this filterPredicate belongs to a WHERE clause
+    }
+
+    @Override
+    public Pair<FilterExpression, FilterExpression> visitAndExpression(final AndFilterExpression expression) {
+        Pair<FilterExpression, FilterExpression> left = expression.getLeft().accept(this);
+        Pair<FilterExpression, FilterExpression> right = expression.getRight().accept(this);
+
+        if (isWhereExpression(left) && isWhereExpression(right)) {
+            // pure-W AND pure-W
+            return Pair.of(
+                    new AndFilterExpression(
+                            extractWhereExpression(left),
+                            extractWhereExpression(right)
+                    ),
+                    null
+            );
+        } else if (isHavingExpression(left) && isWhereExpression(right)) {
+            // pure-H AND pure-W
+            return Pair.of(
+                    extractHavingExpression(left),
+                    extractWhereExpression(right)
+            );
+
+        } else if (isWhereExpression(left) && isHavingExpression(right)) {
+            // pure-W AND pure-H
+            return Pair.of(
+                    extractWhereExpression(left),
+                    extractHavingExpression(right)
+            );
+        } else {
+            return Pair.of(
+                    AndFilterExpression.withLeftAndRight(
```

Comment:
Isn't this the pure having case?

Reply:
This is pure having in some cases. There are also cases when `left` and `right` are both some random mixes of H and W (non-pure)

Reply:
I"m not following.  Pure Having should not have a Where component.